conjure-runtime/src/client_factory.rs

```diff
@@ -401,28 +414,66 @@ impl ClientFactory {
                 move |c| c.merged_service(&service).unwrap_or_default()
             })
             .try_map({
+                let mut client_handles = HashMap::<Host, ClientHandle>::new();
                 move |c| {
+                    let mut new_client_handles = HashMap::new();
                     let mut clients = HashMap::new();
+                    let mut errors = vec![];
 
                     for (i, uri) in c.uris().iter().enumerate() {
                         let host = Host { uri: uri.clone() };
                         let host_config = service_config::Builder::from(c.clone())
                             .uris([uri.clone()])
                             .build();
 
-                        let state = state_builder.build(&host_config, Some(i))?;
-                        let state = Arc::new(ArcSwap::new(state));
-                        let client = make_client(Client::new(state, None));
-
-                        clients.insert(host, client);
+                        let client_handle = match client_handles.remove(&host) {
+                            Some(mut client_handle) => {
+                                if let Err(es) = client_handle.handle.refresh(host_config) {
+                                    errors.extend(es);
+                                }
+                                client_handle
+                            }
+                            None => {
+                                let (host_config, handle) = Refreshable::new(host_config);
+                                let client = match Self::raw_client(
+                                    state_builder.clone(),
+                                    host_config,
+                                    Some(i),
+                                ) {
+                                    Ok(state) => state,
+                                    Err(e) => {
+                                        errors.push(e);
+                                        continue;
+                                    }
+                                };
+                                ClientHandle { client, handle }
+                            }
+                        };
+
+                        clients.insert(host.clone(), make_client(client_handle.client.clone()));
+                        new_client_handles.insert(host, client_handle);
                     }
 
-                    Ok(PerHostClients { clients })
+                    client_handles = new_client_handles;
+                    match errors.pop() {
+                        Some(e) => {
+                            for e2 in errors {
+                                warn!("error reloading per-host clients", error: e2);
+                            }
```

Comment:
The error handling here is all deferred so we don't lose track of state when a subset of clients have issues updating. I don't expect that to be a common case but wanted to make sure we handled it.

Reply:
Where are we updating the state in this method? As in, what would be lost if we returned the error early here https://github.com/palantir/conjure-rust-runtime/pull/219/files#diff-6d2572eabc17f7b3329877aef40ce0daeebbc8cb310e2c5394e530c28b6d42f4R446?

Reply:
We'd forget about some of the clients so the next refresh we'd end up making new ones for those hosts instead of reloading the old ones.